Exception when using --testdox with a dataProvider

| Q | A
| --- | ---
| ParaTest version | 6.9.0
| PHPUnit version | 9.6.4
| PHP version | 8.0.28

#### Summary

Executing a test with a dataprovider and `--testdox` results in an error.

#### Current behavior

This exception is thrown:

```
Fatal error: Uncaught AssertionError: assert(class_exists($class)) in /Users/<user>/Development/<project>/sylius/vendor/brianium/paratest/src/Runners/PHPUnit/ResultPrinter.php:691
Stack trace:
#0 /Users/<user>/Development/<project>/sylius/vendor/brianium/paratest/src/Runners/PHPUnit/ResultPrinter.php(691): assert(false, 'assert(class_ex...')
#1 /Users/<user>/Development/<project>/sylius/vendor/brianium/paratest/src/Runners/PHPUnit/ResultPrinter.php(681): ParaTest\Runners\PHPUnit\ResultPrinter->processTestdoxReader(Object(ParaTest\Logging\JUnit\TestSuite))
#2 /Users/<user>/Development/<project>/sylius/vendor/brianium/paratest/src/Runners/PHPUnit/ResultPrinter.php(302): ParaTest\Runners\PHPUnit\ResultPrinter->processTestdoxReader(Object(ParaTest\Logging\JUnit\TestSuite))
#3 /Users/<user>/Development/<project>/sylius/vendor/brianium/paratest/src/Runners/PHPUnit/Runner.php(108): ParaTest\Runners\PHPUnit\ResultPrinter->printFeedback(Object(ParaTest\Runners\PHPUnit\Suite))
#4 /Users/<user>/Development/<project>/sylius/vendor/brianium/paratest/src/Runners/PHPUnit/Runner.php(41): ParaTest\Runners\PHPUnit\Runner->testIsStillRunning(Object(ParaTest\Runners\PHPUnit\Worker\RunnerWorker))
#5 /Users/<user>/Development/<project>/sylius/vendor/brianium/paratest/src/Runners/PHPUnit/BaseRunner.php(80): ParaTest\Runners\PHPUnit\Runner->doRun()
#6 /Users/<user>/Development/<project>/sylius/vendor/brianium/paratest/src/Console/Commands/ParaTestCommand.php(101): ParaTest\Runners\PHPUnit\BaseRunner->run()
#7 /Users/<user>/Development/<project>/sylius/vendor/symfony/console/Command/Command.php(298): ParaTest\Console\Commands\ParaTestCommand->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#8 /Users/<user>/Development/<project>/sylius/vendor/symfony/console/Application.php(1040): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#9 /Users/<user>/Development/<project>/sylius/vendor/symfony/console/Application.php(301): Symfony\Component\Console\Application->doRunCommand(Object(ParaTest\Console\Commands\ParaTestCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#10 /Users/<user>/Development/<project>/sylius/vendor/symfony/console/Application.php(171): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#11 /Users/<user>/Development/<project>/sylius/vendor/brianium/paratest/bin/paratest(37): Symfony\Component\Console\Application->run()
```

#### How to reproduce: command, code and error stack trace

Create a test with a dataproveder e.g.
```php
use PHPUnit\Framework\TestCase;

class ExampleTest extends TestCase
{
    /**
     * @dataProvider provideTrimData
     */
    public function testTrim($expectedResult, $input): void
    {
        self::assertSame($expectedResult, trim($input));
    }

    public function provideTrimData()
    {
        return [
            'leading space is trimmed' => [
                'Hello World',
                ' Hello World',
            ],
            'trailing space and newline are trimmed' => [
                'Hello World',
                "Hello World \n",
            ],
        ];
    }
}
```

and run it with paratest and testdox:

`php vendor/bin/paratest --testdox ./tests/ExampleTest.php`

#### Expected behavior

Test doesn't throw an exception and works the same as with phpunit or without testdox

--- a/composer.json
+++ b/composer.json
@@ -40,22 +40,22 @@
         "ext-simplexml": "*",
         "fidry/cpu-core-counter": "^0.4.1 || ^0.5.1",
         "jean85/pretty-package-versions": "^2.0.5",
-        "phpunit/php-code-coverage": "^9.2.24",
+        "phpunit/php-code-coverage": "^9.2.25",
         "phpunit/php-file-iterator": "^3.0.6",
         "phpunit/php-timer": "^5.0.3",
-        "phpunit/phpunit": "^9.6.3",
-        "sebastian/environment": "^5.1.4",
-        "symfony/console": "^5.4.16 || ^6.2.5",
-        "symfony/process": "^5.4.11 || ^6.2.5"
+        "phpunit/phpunit": "^9.6.4",
+        "sebastian/environment": "^5.1.5",
+        "symfony/console": "^5.4.21 || ^6.2.7",
+        "symfony/process": "^5.4.21 || ^6.2.7"
     },
     "require-dev": {
         "ext-pcov": "*",
         "ext-posix": "*",
         "doctrine/coding-standard": "^10.0.0",
         "infection/infection": "^0.26.19",
-        "squizlabs/php_codesniffer": "^3.7.1",
-        "symfony/filesystem": "^5.4.13 || ^6.2.5",
-        "vimeo/psalm": "^5.6"
+        "squizlabs/php_codesniffer": "^3.7.2",
+        "symfony/filesystem": "^5.4.21 || ^6.2.7",
+        "vimeo/psalm": "^5.7.7"
     },
     "autoload": {
         "psr-4": {

--- a/src/Runners/PHPUnit/ResultPrinter.php
+++ b/src/Runners/PHPUnit/ResultPrinter.php
@@ -39,6 +39,7 @@ use function fwrite;
 use function get_class;
 use function implode;
 use function is_array;
+use function is_string;
 use function max;
 use function preg_split;
 use function rtrim;
@@ -689,12 +690,14 @@ final class ResultPrinter
 
         $separator = PHP_EOL;
         foreach ($testSuite->cases as $index => $case) {
-            if (0 === $index) {
+            if ($index === 0) {
                 $class = $case->class;
                 assert(class_exists($class));
 
                 $this->output->writeln($prettifier->prettifyTestClass($class));
             }
+
+            assert(isset($class) && is_string($class) && class_exists($class));
 
             $separator = PHP_EOL;
             $testCase  = new $class($case->name);

--- a/src/Runners/PHPUnit/ResultPrinter.php
+++ b/src/Runners/PHPUnit/ResultPrinter.php
@@ -687,13 +687,15 @@ final class ResultPrinter
 
         $prettifier = new NamePrettifier(false);
 
-        $class = $testSuite->name;
-        assert(class_exists($class));
-
-        $this->output->writeln($prettifier->prettifyTestClass($class));
-
         $separator = PHP_EOL;
-        foreach ($testSuite->cases as $case) {
+        foreach ($testSuite->cases as $index => $case) {
+            if (0 === $index) {
+                $class = $case->class;
+                assert(class_exists($class));
+
+                $this->output->writeln($prettifier->prettifyTestClass($class));
+            }
+
             $separator = PHP_EOL;
             $testCase  = new $class($case->name);
             assert($testCase instanceof TestCase);

--- a/test/Unit/Runners/PHPUnit/RunnerTestCase.php
+++ b/test/Unit/Runners/PHPUnit/RunnerTestCase.php
@@ -533,6 +533,19 @@ abstract class RunnerTestCase extends TestBase
         static::assertSame($defaultOrder, $reverseOrderReversed);
     }
 
+    final public function testTestdoxAndDataProvider(): void
+    {
+        $this->bareOptions['--path'] = $this->fixture('testdox_dataprovider');
+        $this->bareOptions['--testdox'] = true;
+
+        $runnerResult = $this->runRunner();
+
+        $output = $runnerResult->getOutput();
+        static::assertStringContainsString('✔ Trim with data set "leading space is trimmed"', $output);
+        static::assertStringContainsString('✔ Trim with data set "trailing space and newline are trimmed"', $output);
+        static::assertStringContainsString('OK (2 tests, 2 assertions)', $output);
+    }
+
     /** @return string[] */
     private function prepareOutputForTestOrderCheck(string $output): array
     {

--- a/test/Unit/Runners/PHPUnit/RunnerTestCase.php
+++ b/test/Unit/Runners/PHPUnit/RunnerTestCase.php
@@ -535,7 +535,7 @@ abstract class RunnerTestCase extends TestBase
 
     final public function testTestdoxAndDataProvider(): void
     {
-        $this->bareOptions['--path'] = $this->fixture('testdox_dataprovider');
+        $this->bareOptions['--path']    = $this->fixture('testdox_dataprovider');
         $this->bareOptions['--testdox'] = true;
 
         $runnerResult = $this->runRunner();

--- a/test/fixtures/testdox_dataprovider/TestdoxDataProviderTest.php
+++ b/test/fixtures/testdox_dataprovider/TestdoxDataProviderTest.php
@@ -1,0 +1,32 @@
+<?php
+
+declare(strict_types=1);
+
+namespace ParaTest\Tests\fixtures\testdox_dataprovider;
+
+use PHPUnit\Framework\TestCase;
+
+final class TestdoxDataProviderTest extends TestCase
+{
+    /**
+     * @dataProvider provideTrimData
+     */
+    public function testTrim(string $expectedResult, string $input): void
+    {
+        self::assertSame($expectedResult, trim($input));
+    }
+
+    public function provideTrimData(): array
+    {
+        return [
+            'leading space is trimmed' => [
+                'Hello World',
+                ' Hello World',
+            ],
+            'trailing space and newline are trimmed' => [
+                'Hello World',
+                "Hello World \n",
+            ],
+        ];
+    }
+}

--- a/test/fixtures/testdox_dataprovider/TestdoxDataProviderTest.php
+++ b/test/fixtures/testdox_dataprovider/TestdoxDataProviderTest.php
@@ -6,16 +6,17 @@ namespace ParaTest\Tests\fixtures\testdox_dataprovider;
 
 use PHPUnit\Framework\TestCase;
 
+use function trim;
+
 final class TestdoxDataProviderTest extends TestCase
 {
-    /**
-     * @dataProvider provideTrimData
-     */
+    /** @dataProvider provideTrimData */
     public function testTrim(string $expectedResult, string $input): void
     {
         self::assertSame($expectedResult, trim($input));
     }
 
+    /** @return array<non-empty-string, list<non-empty-string>> */
     public function provideTrimData(): array
     {
         return [

